File: examples/05-fork-join/Example11.java
// =================================================================
//
// File: Example11.java
// Author(s): Sandra Tello A01703658 Isaac Planter A01702962
// Speeddup: 0.662162162
// Description: This file implements the code that transforms a
//				grayscale image using Java's Fork-Join.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

import java.awt.image.BufferedImage;
import java.io.File;
import javax.imageio.ImageIO;
import java.io.IOException;
import java.util.concurrent.RecursiveAction;
import java.util.concurrent.ForkJoinPool;

public class Example11 extends RecursiveAction {

	private static final int MIN = 15_000;
	private int src[], dest[], width, height, start, end;

	public Example11(int src[], int dest[], int width, int height, int start, int end) {
		this.src = src;
		this.dest = dest;
		this.width = width;
		this.height = height;
		this.start = start;
		this.end = end;
	}

	private void greyPixel(int ren, int col) {
		int  i, j;
		int pixel, dpixel;
		float r, g, b, aux;

		
		r = 0; g = 0; b = 0;
		
		
		pixel = src[(ren * width) + col];

		r += (float) ((pixel & 0x00ff0000) >> 16);
		g += (float) ((pixel & 0x0000ff00) >> 8);
		b += (float) ((pixel & 0x000000ff) >> 0);
		aux = (r+g+b)/3;
		

		dpixel = (0xff000000)
				| (((int) (aux)) << 16)
				| (((int) (aux)) << 8)
				| (((int) (aux)) << 0);
		dest[(ren * width) + col] = dpixel;
	}

	protected void computeDirectly() {
		int index, ren, col;

		for (index = start; index < end; index++) {
			ren = index / width;
			col = index % width;
			greyPixel(ren, col);
		}
	}

	@Override
	protected void compute() {
		if ( (end - start) <= MIN ) {
			computeDirectly();
		} else {
			int mid = start + ((end - start) / 2);
			invokeAll(new Example11(src, dest, width, height, start, mid),
					  new Example11(src, dest, width, height, mid, end));
		}
	}

	public static void main(String args[]) throws Exception {
		long startTime, stopTime;
		double ms;
		int src[], dest[], w, h;
		ForkJoinPool pool;

		if (args.length != 1) {
			System.out.println("usage: java Example11 image_file");
			System.exit(-1);
		}

		final String fileName = args[0];
		File srcFile = new File(fileName);
        final BufferedImage source = ImageIO.read(srcFile);

		w = source.getWidth();
		h = source.getHeight();
		src = source.getRGB(0, 0, w, h, null, 0, w);
		dest = new int[src.length];

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();

			pool = new ForkJoinPool(Utils.MAXTHREADS);
			pool.invoke(new Example11(src, dest, w, h, 0, (w * h)));

			stopTime = System.currentTimeMillis();
			ms += (stopTime - startTime);
		}

		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
		final BufferedImage destination = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
		destination.setRGB(0, 0, w, h, dest, 0, w);

		

		try {
			ImageIO.write(destination, "png", new File("grey.png"));
			System.out.println("Image was written succesfully.");
		} catch (IOException ioe) {
			System.out.println("Exception occured :" + ioe.getMessage());
		}
	}
	// place your code here
}

File: examples/07-tbb/example11.cpp
// =================================================================
//
// File: example11.cpp
// Author(s): Sandra Tello A01703658 Isaac Planter A01702962
// Speedup: 3.464595989
// Description: This file implements the code that transforms a
//				grayscale image using Intel's TBB. Uses OpenCV,
//				to compile:
//	g++ example11.cpp `pkg-config --cflags --libs opencv4` -ltbb
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//.
// =================================================================

#include <iostream>
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/highgui/highgui_c.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/imgcodecs/imgcodecs.hpp>
#include <tbb/parallel_for.h>
#include <tbb/blocked_range.h>
#include "utils.h"

using namespace std;
using namespace cv;
using namespace tbb;

class Grayscale {
private:
	Mat &src, &dest;

	void grayPixel(int ren, int col) const {
		float r, g, b, aux;

		r = 0; g = 0; b = 0;

		r = (float) src.at<cv::Vec3b>(ren, col)[RED];
		g = (float) src.at<cv::Vec3b>(ren, col)[GREEN];
		b = (float) src.at<cv::Vec3b>(ren, col)[BLUE];
		aux = (r+g+b)/3;

		dest.at<cv::Vec3b>(ren, col)[RED] =  (unsigned char) (aux);
		dest.at<cv::Vec3b>(ren, col)[GREEN] = (unsigned char) (aux);
		dest.at<cv::Vec3b>(ren, col)[BLUE] = (unsigned char) (aux);
	}

public:
	Grayscale(Mat &s, Mat &d) : src(s), dest(d) {}

	void operator() (const blocked_range<int> &r) const {
		for (int i = r.begin(); i != r.end(); i++) {
			for(int j = 0; j < src.cols; j++) {
				grayPixel(i, j);
			}
		}
	}
};

int main(int argc, char* argv[]) {
	double ms;

	if (argc != 2) {
		printf("usage: %s source_file\n", argv[0]);
		return -1;
	}

	Mat src = imread(argv[1], cv::IMREAD_COLOR);
	Mat dest = Mat(src.rows, src.cols, CV_8UC3);
	if (!src.data) {
		printf("Could not load image file: %s\n", argv[1]);
		return -1;
	}

	cout << "Starting..." << endl;
	ms = 0;
	for (int  i = 0; i < N; i++) {
		start_timer();

		// place your code here
		Grayscale obj(src, dest);
		parallel_for(blocked_range<int>(0, src.rows),  obj);
		ms += stop_timer();
	}

	cout << "avg time = " << setprecision(15) << (ms / N) << " ms" << endl;

	/*
	cv::namedWindow("Original", cv::WINDOW_AUTOSIZE);
	cv::imshow("Original", src);

	cv::namedWindow("Gray scale", cv::WINDOW_AUTOSIZE);
	cv::imshow("Gray scale", dest);

	cv::waitKey(0);
	*/

	cv::imwrite("gray_scale.png", dest);

	return 0;
}
